src/webui/www/private/scripts/dynamicTable.js

```diff
@@ -831,64 +840,152 @@ window.qBittorrent.DynamicTable ??= (() => {
+                // rerender on table update
+                this.rerender();
+            }
+            else {
+                const trs = [...this.getTrs()];
 
-            for (let rowPos = 0; rowPos < rows.length; ++rowPos) {
-                const rowId = rows[rowPos]["rowId"];
-                let tr_found = false;
-                for (let j = rowPos; j < trs.length; ++j) {
-                    if (trs[j]["rowId"] === rowId) {
-                        tr_found = true;
-                        if (rowPos === j)
+                for (let rowPos = 0; rowPos < rows.length; ++rowPos) {
+                    const rowId = rows[rowPos]["rowId"];
+                    let tr_found = false;
+                    for (let j = rowPos; j < trs.length; ++j) {
+                        if (trs[j]["rowId"] === rowId) {
+                            tr_found = true;
+                            if (rowPos === j)
+                                break;
+                            trs[j].inject(trs[rowPos], "before");
+                            const tmpTr = trs[j];
+                            trs.splice(j, 1);
+                            trs.splice(rowPos, 0, tmpTr);
                             break;
-                        trs[j].inject(trs[rowPos], "before");
-                        const tmpTr = trs[j];
-                        trs.splice(j, 1);
-                        trs.splice(rowPos, 0, tmpTr);
-                        break;
+                        }
+                    }
+                    if (tr_found) { // row already exists in the table
+                        this.updateRow(trs[rowPos], fullUpdate);
+                    }
+                    else { // else create a new row in the table
+                        const tr = this.createRowElement(rows[rowPos]);
+
+                        // Insert
+                        if (rowPos >= trs.length) {
+                            tr.inject(this.tableBody);
+                            trs.push(tr);
+                        }
+                        else {
+                            tr.inject(trs[rowPos], "before");
+                            trs.splice(rowPos, 0, tr);
+                        }
+
+                        // Update context menu
+                        this.contextMenu?.addTarget(tr);
+
+                        this.updateRow(tr, true);
                     }
                 }
-                if (tr_found) { // row already exists in the table
-                    this.updateRow(trs[rowPos], fullUpdate);
+
+                const rowPos = rows.length;
+
+                while ((rowPos < trs.length) && (trs.length > 0))
+                    trs.pop().destroy();
+            }
+        },
+
+        rerender: function() {
+            if (!this.useVirtualList)
+                return;
+            const rows = this.getFilteredAndSortedRows();
+
+            for (let i = 0; i < this.selectedRows.length; ++i) {
+                if (!(this.selectedRows[i] in rows)) {
```

Comment:
For a second I thought I was reading python code...
Anyway, use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

Reply:
It does not work in the first place.

```js
"foo" in ["foo", "bar"] // false 
```

Reply:
>this is copy pasted old code, needs refactoring?

For existing code, you are not required to change it. Your choice.
But personally I would change it if the scope is limited and isn't too complex.

Reply:
It works in a wrong way. MooTools can't override such core language operators.

This check is pointless by its own. JS `in` operator checks for **keys** in **objects**. But for arrays indexes became keys. I.e.
```js
1 in ["foo", "bar"] // true
```
because
```js
["foo", "bar"]
/*
Array [ "foo", "bar" ]
	0: "foo"
	1: "bar"
	length: 2
*/
```

So `this.selectedRows[i] in rows` simply works as `this.selectedRows[i] < rows.length` in this case.
Which I'm not even sure being right logically.

Reply:
Well, that's completely messed up and requires a full rewrite.
But you've already got rid of that part anyway, so yeah.